crates/topos-p2p/src/behaviour/grpc.rs

```diff
@@ -0,0 +1,516 @@
+            id: connection_id,
+            address,
+            request_id: None,
+            channel: None,
+        };
+
+        self.connected.entry(peer_id).or_default().push(connection);
+
+        // TODO refactor this to better handle connection
+        // If there is no current established connection it means that it's the
+        // first connection with that peer
+        if other_established == 0 {
+            self.try_connect(&peer_id);
+        }
+    }
+
+    /// Starts the gRPC service if not already started
+    fn on_new_listener(&mut self, listener_id: ListenerId) {
+        if let Some(service) = self.service.take() {
+            let (tx, rx) = mpsc::unbounded_channel();
+            self.inbound_stream = Some(tx);
+            // TODO: Switch to serve_with_incoming_shutdown at some point
+            tokio::spawn(service.serve_with_incoming(proxy::GrpcProxy::new(rx)));
+            info!("New gRPC proxy start and listening on {listener_id:?}");
+        } else {
+            warn!("Tried to instantiate gRPC proxy but service is missing");
+        }
+    }
+
+    /// On [`ConnectionClosed`] we cleanup the `connected` state of the behaviour.
+    fn on_connection_closed(
+        &mut self,
+        ConnectionClosed {
+            peer_id,
+            connection_id,
+            endpoint,
+            handler,
+            remaining_established,
+        }: ConnectionClosed<<Self as NetworkBehaviour>::ConnectionHandler>,
+    ) {
+        warn!("Connection {connection_id} closed with peer {peer_id}");
+        if let Some(connections) = self.connected.get_mut(&peer_id) {
+            connections.retain(|conn| conn.id != connection_id);
+            if connections.is_empty() {
+                self.connected.remove(&peer_id);
+            }
+        }
+    }
+
+    /// Handle the [`DialFailure`] event comming from the [`Swarm`]
+    fn on_dial_failure(
+        &mut self,
+        DialFailure {
+            peer_id,
+            error,
+            connection_id,
+        }: DialFailure,
+    ) {
+        println!("Dial failure {:?}", error);
+        if let Some(peer_id) = peer_id {
+            match error {
+                DialError::DialPeerConditionFalse(_) => {
+                    self.try_connect(&peer_id);
+                }
+                _ => {
+                    if let Some(OutboundConnectionRequest {
+                        request_id,
+                        notifier,
+                    }) = self.pending_outbound_connections.remove(&peer_id)
+                    {
+                        self.pending_events.push_back(ToSwarm::GenerateEvent(
+                            Event::OutboundFailure {
+                                peer_id,
+                                request_id,
+                                error: OutboundError::DialFailure,
+                            },
+                        ));
+
+                        let _ = notifier.send(Err(OutboundError::DialFailure));
```

Comment:
It looks like we're loose track of what the error causing the dial failure. I'm not what happens next here, but whatever code is handling `Event::OutboundFailure` and `OutboundError::DialFailure` will not have access to the error. Is that what we want?

Reply:
We don't really care about the reason of this `DialFailure`, we only care about the `condition` failure.

Reply:
🤷‍♂️ Fair enough.